djautotask/models.py

```diff
@@ -129,22 +160,35 @@ def save(self, *args, **kwargs):
 
         If update_at as a kwarg is True, then update Autotask with changes.
         """
-
+        changed_fields = kwargs.pop('changed_fields', None)
         update_at = kwargs.pop('update_at', False)
-        super().save(*args, **kwargs)
-        if update_at:
-            self.update_at()
 
-    def update_at(self, data=None):
-        if data:
-            fields_to_update = {}
-            for field, data in data.items():
-                fields_to_update[field] = data
-        else:
-            fields_to_update = {
-                'Status': self.status.id,
-            }
-        return api.update_object('Ticket', self.id, fields_to_update)
+        if update_at and changed_fields:
+            self.update_at(changed_fields=changed_fields)
+
+        super().save(**kwargs)
+
+    def update_at(self, **kwargs):
+        api_client = api_rest.TicketsAPIClient()
+        changed_fields = kwargs.get('changed_fields')
+        updated_objects = self.get_changed_values(changed_fields)
+
+        return api_client.update_ticket(self, updated_objects)
+
+    def update_resource(self, extra=None):
```

Comment:
My plan is to return `update_resource` to `ResourceAssignableModel` after Task REST API because, for now, both `update_resource` methods cannot be compatible.

So, it will be easier to remember to put this method here (a little bit closer to `ResourceAssignableModel`).  :)

Reply:
Ok, that is reasonable. 

But I'm thinking we can probably get rid of `ResourceAssignableModel` once the move to REST is complete (at least for Ticket and Task). I think it was more acceptable to have that with the SOAP architecture but with REST it makes a lot more sense to me to have those update methods in the API client class. Especially since it already has an update_ticket method.

Reply:
For the code:
```python
        if extra:
            try:
                keys = []
                for key in extra.keys():
                    key = key[0].lower() + key[1:]
                    keys.append(key)
                changed_fields += keys
            except TypeError:
                changed_fields += extra
```
Based on your TODO this will be improved/updated when Task moves to REST. But maybe the variable names could be made all little less generic? I find it hard to read and Task and Ticket are quite similar so I don't think it would need to be super generic.

Reply:
But, here `Resource` means AT resource owning ticket, so it seems not be general to be put into the API client class. 

So do you mean `TicketsAPIClient`? It seems also uncomfortable because its job including calling update_at, which also asks API client to complete the job.

So we can regard `update_resource` as model update method (only for resource related fields), and it seems fine here.

Most of all, in our Kanban app, Ticket cards have `update_resource` method, so it seems the best approach to get along together with our other apps.

Reply:
Good idea! I will do.

Reply:
You make good points but I'm still not sure. 

@kti-sam can we get a second look and opinion on this from you?

Reply:
Yeah, this isn't ideal. If I am understanding correctly though we don't need to put this in the API or even keep the `update_resource` method here on the model. Updating the resource should be able to just use `update_at`, correct?

If I'm not wrong about that it looks like the `TicketResourceMixin` and `TaskResourceMixin` already have their own methods for how to update resources in `manage_resources.py`. `TaskResourceMixin.perform_update`  can mostly stay the same and `TaskResourceMixin.perform_update` can call `update_at` in a way that is compatible with the REST API.

Then the only thing left is for this call to `update_resource` around line 470 in `manage_resources.py`:
```
    def _drag_assignment(
            self, card, old_resource, old_resource_role, old_status):
        card.update_resource({'Status': card.status.id})
        card.save()
```
I'm pretty sure that can be changed to use `perform_update`. 

With that done, all the task vs. ticket logic remains in `manage_resources.py`, and we don't need any temporary hacks and TODOs in `django-autotask`. Also I think that make `update_resource` completely unused as Tasks already just use `update_at` in `perform_update`, so maybe it can be removed entirely? (not counting where tasks use `update_resource` on line 470, as mentioned that can be replaced with `perform_update` I think)

This was just from me eyeballing this quickly, so I'll leave you to double check my math on that Cameron and Daniel :laughing:

Reply:
Thank Cameron for raising this issue, and thanks Sam to present big picture. I will prepare another version. Just a min.

Reply:
Thanks for the comprehensive answer. 

That sounds like a good plan to me. Yeah, updating resources should be able to use `update_at` as far as I can tell. It's just updating a field on a ticket. And I'm pretty sure `update_resource` and its mixin can be removed as you describe.

Daniel, does that seem reasonable to you?

Reply:
Yes, reasonable and better approach!

Reply:
Finally updated and new issue is registered. The following issue happens in current master branch. Thanks.

https://git.ops.topleft.team/apps/django-topleft/-/issues/2037

Reply:
Thanks for finding and opening that new issue. 

Your changes look good. Merging shortly.